### PR TITLE
Implement EeGenktAlgorithm

### PR DESCRIPTION
**Pull request by Bastiantheone**
How often is the go-hep.org website updated with the GitHub code? Since this code needs the new CosTheta method from fmom, should I leave the import statement as it is, or should I import it from the GitHub code?

### CODE REVIEW

**Comment by sbinet** [comment]
it should be instantaneous.

what `go-hep.org/x/hep` does is actually redirect to `github.com/go-hep/hep`.

see:

```
$> rm -rf $GOPATH/src/go-hep.org/x/hep
$> go get -v go-hep.org/x/hep
Fetching https://go-hep.org/x/hep?go-get=1
Parsing meta tags from https://go-hep.org/x/hep?go-get=1 (status code 200)
get "go-hep.org/x/hep": found meta tag main.metaImport{Prefix:"go-hep.org/x/hep", VCS:"git", RepoRoot:"https://github.com/go-hep/hep"} at https://go-hep.org/x/hep?go-get=1
go-hep.org/x/hep (download)
go-hep.org/x/hep
```

**Comment by sbinet** [comment]
the travis failure comes from some non-gofmt'ed files.

please run:

```
$> go get -u golang.org/x/tools/cmd/goimports
$> cd $GOPATH/src/go-hep.org/x/hep
$> goimports -w .
```

and recommit.

**Comment by Bastiantheone** [comment]
Okay now it passed :) I'll remember to run goimports next time

File: fastjet/clustersequence.go [hunk 1]
```diff
@@ -360,11 +370,30 @@ func (cs *ClusterSequence) runN3Dumb() error {
...
+					y := math.MaxFloat64
+					if den != 0 {
+						y = jetscale * (1 - fmom.CosTheta(&ijet.PxPyPzE, &jjet.PxPyPzE)) / den
+					}
+					if y < ymin {
+						ymin = y
+						ii = i
+						jj = j
+					}
+					break
```
Comment (sbinet) [inline fastjet/clustersequence.go]: Unlike `C`, the various `case` blocks of a `switch` control structure do not need any `break` in `Go`.
see:
 https://golang.org/ref/spec#Switch_statements

```
In an expression switch, the switch expression is evaluated and the case expressions, which need not be constants, are evaluated left-to-right and top-to-bottom; the first one that equals the switch expression triggers execution of the statements of the associated case; the other cases are skipped.
```

just remove that `break`

File: fastjet/clustersequence.go [hunk 2]
```diff
@@ -360,11 +370,30 @@ func (cs *ClusterSequence) runN3Dumb() error {
...
+				default:
+					y := jetscale * Distance(ijet, jjet) * cs.invR2
+					if y < ymin {
+						ymin = y
+						ii = i
+						jj = j
+					}
+					break
```
Comment (sbinet) [inline fastjet/clustersequence.go]: same here: just remove the `break`.

File: fastjet/clustersequence.go [hunk 3]
```diff
@@ -360,11 +370,30 @@ func (cs *ClusterSequence) runN3Dumb() error {
...
+				default:
+					y := jetscale * Distance(ijet, jjet) * cs.invR2
+					if y < ymin {
+						ymin = y
+						ii = i
+						jj = j
+					}
```
Comment (sbinet) [inline fastjet/clustersequence.go]: the whole if-block is repeated for each of the 2 cases.

couldn't we declare the `y` variable before the `switch`, fill it with the correct value inside the `case` clauses, and move the `if y < ymin { ... }` block after the `switch` ?

**Comment by sbinet** [comment]
wrt `goimports`: depending on your favorite editor, there might be a way to automatically hook it up and run it whenever you hit save.

see:
 https://github.com/golang/go/wiki/IDEsAndTextEditorPlugins

with `goimports` running whenever you hit save, you won't have to add/remove _e.g._ `import "fmt"` when you go through a little `fmt.Printf("foo\n")` debugging session :)

**Comment by sbinet** [comment]
the one commit should probably read the following as a commit message:

```
fastjet: implement EeGenKtAlgorithm

This CL implements the e+e- general kt algorithm for fastjet.
This CL also enables the correspondig test for that algorithm.
```

If you don't know how to merge multiple commits into one, here is a little tutorial.

Assuming you have the following setup:

```
$> cd $GOPATH/src/go-hep.org/x/hep
$> git remote -v
origin	https://github.com/Bastiantheone/hep.git (fetch)
origin	https://github.com/Bastiantheone/hep.git (push)
upstream	https://github.com/go-hep/hep.git (fetch)
upstream	https://github.com/go-hep/hep.git (push)
```

then, the following should work:

```sh
## go to the PR branch (eegenktalgo)
$> git checkout eegenktalgo

## retrieve the latest updates from upstream
$> go fetch --all -p

## display the current status
$> git log --oneline --graph --decorate --all | head
* 38c5544 (upstream/master, upstream/HEAD) hep: display output of gofmt/goimports when command fails
| * 04aa131 (HEAD -> eegenktalgo, origin/eegenktalgo) fastjet: clean up code
| * e4deaf2 fastjet: run goimports
| * c24fea2 fastjet_test: enable testing for EeGenKtAlgorithm
| * 2161422 fastjet: implement EeGenKtAlgorithm
| | * 97c1346 (upstream/rootio-ttree-reader) rootio: first stab at evnt-tree
| |/  
|/|   
* | 0b78c29 hep: add a test for gofmtd code
* | 56206eb lhef: add copyright headers, remove spurious printouts
* | 26d4b11 fmom,lhef,rootio: apply gofmt
|/  
* a964065 (origin/master) fmom: add Testing for Cos(Theta)
* 93cdba4 fmom: add Cos(Theta) and Dot Product between 2 four-vectors
* 9ff8704 hep: add DOI badge
* a17e090 (tag: v0.1) rootio/cmd/root-gen-datareader: improve documentation
```

we squash all 4 commits (`04aa131`, `e4deaf2`, `c24fea2` and `2161422`) into one, on top of `origin/master`:

```
$> git rebase -i origin/master
```

This should open your favorite `$EDITOR` with the following content:

```
pick 2161422 fastjet: implement EeGenKtAlgorithm
pick c24fea2 fastjet_test: enable testing for EeGenKtAlgorithm
pick e4deaf2 fastjet: run goimports
pick 04aa131 fastjet: clean up code

# Rebase a964065..04aa131 onto a964065 (4 commands)
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
# d, drop = remove commit
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out
```

You should modify like so:

```
r 2161422 fastjet: implement EeGenKtAlgorithm
f c24fea2 fastjet_test: enable testing for EeGenKtAlgorithm
f e4deaf2 fastjet: run goimports
f 04aa131 fastjet: clean up code
```

so you can then modify the commit message of `2161422` like I suggested above.

and voila:

```sh
## [...]
## output of the 'git rebase -i origin/master' command
git rebase -i origin/master
[detached HEAD 2e8bd0d] fastjet: implement EeGenKtAlgorithm
 Author: Bastian Wieck <b.wieck@gmx.de>
 Date: Thu Apr 6 13:25:34 2017 -0400
 1 file changed, 33 insertions(+), 5 deletions(-)
[detached HEAD 4abdd17] fastjet: implement EeGenKtAlgorithm
 Author: Bastian Wieck <b.wieck@gmx.de>
 Date: Thu Apr 6 13:25:34 2017 -0400
 2 files changed, 23 insertions(+), 5 deletions(-)
Successfully rebased and updated refs/heads/eegenktalgo.

## check that everything is as we want
$> git log --oneline --graph --decorate --all | head
* 4abdd17 (HEAD -> eegenktalgo) fastjet: implement EeGenKtAlgorithm
| * 38c5544 (upstream/master, upstream/HEAD) hep: display output of gofmt/goimports when command fails
| | * 04aa131 (origin/eegenktalgo) fastjet: clean up code
| | * e4deaf2 fastjet: run goimports
| | * c24fea2 fastjet_test: enable testing for EeGenKtAlgorithm
| | * 2161422 fastjet: implement EeGenKtAlgorithm
| |/  
|/|   
| | * 97c1346 (upstream/rootio-ttree-reader) rootio: first stab at evnt-tree
| |/  
| * 0b78c29 hep: add a test for gofmtd code
| * 56206eb lhef: add copyright headers, remove spurious printouts
| * 26d4b11 fmom,lhef,rootio: apply gofmt
|/  
* a964065 (origin/master) fmom: add Testing for Cos(Theta)
* 93cdba4 fmom: add Cos(Theta) and Dot Product between 2 four-vectors
* 9ff8704 hep: add DOI badge
* a17e090 (tag: v0.1) rootio/cmd/root-gen-datareader: improve documentation
* 24580d8 rootio/cmd/root-srv: improve documentation

## (force) push the new eegenktalgo branch
$> git push -f origin eegenktalgo
```

congratulations, you've reached `fair-git-ninja` level :)
(and you also made the git history of `go-hep/hep` slightly better and more readable in the process)

if you didn't manage or have some issue, just let me know and we'll sort this out.

**Comment by sbinet** [comment]
merged w/ 703f167

(apologies for the belated merge, it seems I wasn't notified by github that you pushed a new revision...)

**Pull request closed without merging**